### PR TITLE
server: add disaggregated prefill via HTTP state handoff

### PR DESCRIPTION
**Pull request by ServeurpersoCom**
## Overview

A server started with --prefill-url delegates prompt processing of requests that set prefill: true to a remote server running the same model. The prompt is forwarded with an x-prefill header, the remote returns the serialized sequence state and tokens as a binary body, and the state is applied to the slot when the task launches so the regular prompt cache path reuses it.

The prefill server stops one token early: recurrent state in hybrid models only exists at the last processed position, so shipping the state at n-1 lets the decode server compute the final token itself, obtaining the logits and advancing any recurrent state.

## Additional information

POC https://github.com/ggml-org/llama.cpp/issues/21266 based on @ngxson idea to simplify

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES

### CODE REVIEW

**Comment by 200lz** [comment]
Since the serialized sequence state now crosses a process/server boundary, would it be useful to bind the payload to a small compatibility identity — model/tokenizer identity plus llama.cpp/state-format version — and reject mismatches before applying it? That could turn “both servers run the same model” from an operational assumption into a validated handoff contract.

**Comment by ServeurpersoCom** [comment]
Agreed, good v1 hardening, the state format itself is already versioned, the model identity check is the missing piece, keeping the POC minimal until the approach settles

**Comment by 200lz** [comment]
That makes sense—keeping the POC minimal while the interface settles sounds right. Once it moves toward hardening, I’d be happy to help sketch a small negative-test matrix for model-identity mismatches.

**Comment by ServeurpersoCom** [comment]
I'm working half blind here until my Mac arrives, both my GPUs are fast enough that disagg overhead dominates, so any numbers from a real asymmetric setup would help a lot. I like seeing actual gains in practice, not just a mechanism that works:)

**Comment by 200lz** [comment]
That sounds like a useful test. I can look into setting up a genuinely asymmetric configuration and collecting reproducible numbers.

Would a faster prefill GPU paired with a lower-tier decode GPU over low-latency networking be representative? I’m thinking of comparing local-only vs disaggregated TTFT across several prompt lengths, while reporting state size, handoff time, exact build/model identity, and deterministic output parity. Happy to adjust the model or topology to what would be most useful for this PR.

**Comment by ServeurpersoCom** [comment]
That topology works: what matters for TTFT is the prefill compute ratio between the nodes versus the transfer cost. Two things: use fresh prompts or cold slots for the local-only baseline, the prompt cache flatters it otherwise, and sweep prompt length into the tens of thousands. Bonus if you can compare a pure attention vs a hybrid/SWA model, their state sizes differ a lot. Parity: greedy decoding, compare full output strings.

File: tools/server/server-context.cpp [hunk 1]
```diff
@@ -39,6 +41,43 @@ using json = nlohmann::ordered_json;
 
 constexpr int HTTP_POLLING_SECONDS = 1;
 
+// wire format of a serialized prefill state: magic, token count, tokens, raw sequence state
+constexpr uint32_t PREFILL_STATE_MAGIC           = 0x50524631; // "PRF1"
+constexpr int      PREFILL_CONNECT_TIMEOUT_SECS  = 10;
+constexpr int      PREFILL_RESPONSE_TIMEOUT_SECS = 3600;
+
+static std::string prefill_state_pack(const llama_tokens & tokens, const std::vector<uint8_t> & state) {
```
Comment (ngxson) [inline tools/server/server-context.cpp]: just wondering if we could just use the `server_tokens::deserialize` / `server_tokens::serialize` interface; it also covered the multimodal path

Comment (ngxson) [reply tools/server/server-context.cpp]: ah nvm, yes we can; indeed we can just abstract the handling of `SERVER_TASK_TYPE_SLOT_SAVE` / `SERVER_TASK_TYPE_SLOT_RESTORE` to a global func

let's do that later on, not very important for now

Comment (ServeurpersoCom) [reply tools/server/server-context.cpp]: It allows for broader testing! Done, the wire now carries server_tokens::serialize() output, which brings the version check and the multimodal path for free. Left the SLOT_SAVE/RESTORE global func abstraction for the follow-up as you said.

**Comment by 200lz** [comment]
<html>
<body>
<html><head></head><body><p>I followed up with the requested asymmetric numbers on the PR's exact current head (<code inline="">5bd8a22307d9963d9bea4f1d588eef6352a070cf</code>).</p><h3>Setup</h3><ul><li><p>prefill: Tesla V100-SXM2-32GB</p></li><li><p>decode: Tesla T4 16GB</p></li><li><p>same Tencent Tokyo VPC/subnet: ~0.14 ms ping; ~3.2 Gbit/s iperf3</p></li><li><p>Ubuntu 22.04; NVIDIA driver 580.126.20; CUDA toolkit 12.9</p></li><li><p>pure-attention Qwen2.5-7B-Instruct Q4_K_M</p></li><li><p><code inline="">ctx-size=32768</code>, <code inline="">parallel=1</code>, f16 KV, flash attention off</p></li><li><p>fresh unique prompts, <code inline="">cache_prompt=false</code>, greedy decoding</p></li><li><p>model SHA256: <code inline="">1875fb29e8c91c86615c00e92d8b4114e56bc24359adb5a8db8b36452fae4a49</code></p></li></ul><h3>Exact-head boundary</h3><p>The unmodified PR head works below a serialized response size of 100 MiB, but fails immediately above it:</p><ul><li><p>1,800 prompt tokens -&gt; 1,799 state tokens, 98.404 MiB: pass with full-output parity</p></li><li><p>1,850 prompt tokens -&gt; 1,849 state tokens, 101.139 MiB: decode endpoint returns HTTP 400, <code inline="">prefill server request failed, status = -1</code></p></li><li><p>2,048 prompt tokens -&gt; 2,047 state tokens, 111.969 MiB: same failure</p></li></ul><p>The V100 produced both larger states. Both servers remained healthy, with no OOM, XID, kernel error, or process exit.</p><p>This boundary is consistent with the vendored cpp-httplib default <code inline="">CPPHTTPLIB_PAYLOAD_MAX_LENGTH</code> of 100 MiB. The prefill client does not currently override that response limit, and a falsy <code inline="">Result</code> is collapsed to status <code inline="">-1</code>.</p><h3>Diagnostic rerun</h3><p>For diagnosis only, I applied a 13-line test patch that:</p><ol><li><p>calls <code inline="">set_payload_max_length(4 GiB)</code> on the prefill client;</p></li><li><p>reports <code inline="">httplib::to_string(r.error())</code> when no response is returned.</p></li></ol>Diagnostic patch attachment:

https://github.com/user-attachments/files/31062179/payload4g-test-only.patch

Patch SHA256:

`8c0ffed9aa364400035e6be16a1c8b4a2de85295e284e70eea7b38cd9ae70704`

<p>Patched server SHA256:</p><p><code inline="">9d8839290977047bb6ebd6b46714cd469e0fa86948f0a3dadb2407e0c5b18973</code></p><p>With that diagnostic patch, all 15/15 runs passed through 30,000 prompt tokens:</p>
With that diagnostic patch, all 15/15 runs passed through 30,000 prompt tokens:

- 512 tokens: local TTFT 526.71 ms; disaggregated TTFT 1087.59 ms; ratio 2.024x; state 27.952 MiB; receive 601.72 ms.
- 2,048 tokens: local TTFT 2049.30 ms; disaggregated TTFT 3635.34 ms; ratio 1.785x; state 111.969 MiB; receive 1663.01 ms.
- 8,192 tokens: local TTFT 11249.46 ms; disaggregated TTFT 18171.83 ms; ratio 1.615x; state 448.040 MiB; receive 7059.72 ms.
- 16,384 tokens: local TTFT 32784.64 ms; disaggregated TTFT 49425.74 ms; ratio 1.508x; state 896.133 MiB; receive 17299.63 ms.
- 30,000 tokens: local TTFT 99202.73 ms; disaggregated TTFT 139918.17 ms; ratio 1.410x; state 1640.914 MiB; receive 43579.49 ms.



<p>Each target used three repetitions. All 15 runs had exact prompt length, <code inline="">cache_n=0</code>, and byte-identical full greedy output between local and disaggregated execution.</p><p>So state portability and correctness looked good after lifting the HTTP limit, but this V100 -&gt; T4 setup did not show a TTFT win. Serialized-state transfer plus receive/apply overhead dominated end-to-end latency; at 30k, the 1.60 GiB state took 43.58 seconds median just to receive.</p><p>Potential follow-ups seem to be:</p><ul><li><p>configurable response limit and explicit cpp-httplib error reporting;</p></li><li><p>chunked or streamed state transfer;</p></li><li><p>compression or a more compact transferable state representation.</p></li></ul><p>Scope note: this run covered the requested pure-attention sweep only; I did not run the hybrid/SWA bonus comparison.</p><p>I preserved the exact-head failure evidence, diagnostic patch, raw SSE, exact prompt-token hashes, output hashes, both server logs, and the complete 15-run result set. The locally verified evidence archive SHA256 is <code inline="">b01b36241d5179f45c0801c484601da890f77b36bbd29c06a39c2f95e55afc59</code>.</p></body></html>
</body>
</html>
